src/aspire/denoise/denoise_cov2d.py

```diff
@@ -0,0 +1,87 @@
+import logging
+import numpy as np
+
+from aspire.denoise import Denoiser
+from aspire.utils.filters import RadialCTFFilter
+from aspire.utils.blk_diag_func import blk_diag_partition
+from aspire.utils.blk_diag_func import blk_diag_eye
+from aspire.basis.ffb_2d import FFBBasis2D
+from aspire.estimation.covar2d import RotCov2D
+
+logger = logging.getLogger(__name__)
+
+
+class DenoiserCov2D(Denoiser):
+    """
+    Define a derived class for denoising 2D images using Cov2D method
+    """
+    def __init__(self, src, basis, var_noise, ctf_info):
+        """
+        Initialize an object for denoising 2D images using Cov2D method
+
+        :param src: The source object of 2D images with metadata
+        :param basis: The basis method to expand 2D images
+        :param var_noise: The estimated variance of noise
+        :param ctf_info: Whether the CTF filters are included
+        """
+        super().__init__(src)
+        self.imgs_noise = self.src.images(start=0, num=np.inf).asnumpy()
+        self.var_noise = var_noise
+        if not isinstance(basis, FFBBasis2D):
+            raise NotImplementedError('Currently only fast FB method is supported')
+        self.coeffs_noise = None
+        self.coeffs_estim = None
+        self.ctf_idx = None
+        self.ctf_fb = None
+        self.basis = basis
+
+        # Assign the CTF information and index for each image
+        if ctf_info and not (src.filters is None):
+            logger.info(f'Convert non radial CTF filters to radial ones')
+            uniq_rad_filters = []
+            uniq_ctf_filters = list(set(src.filters))
+            # Note that the reason we create the new radial filter list is that the src.filters
+            # have be converted to MultiplicativeFilter class.
+            for f in uniq_ctf_filters:
+                filt = RadialCTFFilter()
+                filt.convert_from(f)
```

Comment:
I thought about it before the implementation. For the first case, it is difficult for Python to have multiple constructor.  For the second case the original filters have been converted to other type and the evaluation function is not available. Anyway we need to reconsider the current implementation of applying filters. It might be more reasonable to keep a list of different filters and apply them later with options instead at the step of loading images. I

Reply:
Move the convert function to Non-radial CTF class and reorganize the code in commit  1693ad8.

Reply:
Implement the `fb_mat` function to generate the FB representation for general non-radial filters in commit 4a18f0e.